### ISSUE TITLE
'unknown ca' with https://data.mixpanel.com

### ISSUE THREAD

**Issue by hamann** [benoitc/hackney#215]
hackney: 1.2.0
ssl_verify_hostname: 1.0.5
otp: 17.5

We cannot make requests to https://data.mixpanel.com with hackney. Browsers don't have any problems with the certificates (chrome,ff,safari..)

```elixir
iex(3)> :hackney.request(:get, <<"https://data.mixpanel.com">>,[],<<>>,[])
[error] SSL: :certify: ssl_handshake.erl:1401:Fatal error: unknown ca

{:error, {:tls_alert, 'unknown ca‘}}
```

whereas https://mixpanel.com works
```elixir
iex(2)> :hackney.request(:get, <<"https://mixpanel.com">>,[],<<>>,[])
{:ok, 200,
 [{"Server", "nginx/1.1.19"}, {"Date", "Fri, 17 Jul 2015 09:01:46 GMT"}, {"Content-Type", "text/html; charset=utf-8"},
  {"Transfer-Encoding", "chunked"}, {"Connection", "keep-alive"}, {"Vary", "Accept-Encoding"}, {"Vary", "Cookie"},
  {"X-Frame-Options", "SAMEORIGIN"},
  {"Set-Cookie",
   "mp__origin_referrer=; Domain=.mixpanel.com; expires=Mon, 11-Jul-2016 09:01:46 GMT; Max-Age=31104000; Path=/"},
  {"Set-Cookie", "mp__origin=; Domain=.mixpanel.com; expires=Mon, 11-Jul-2016 09:01:46 GMT; Max-Age=31104000; Path=/"}],
 #Reference<0.0.0.4580>}
```

The difference between the two domains is that data.mixpanel.com is providing a (wrong?) root certificate. Maybe that's the problem here?

ssllabs tests:
https://www.ssllabs.com/ssltest/analyze.html?d=data.mixpanel.com
https://www.ssllabs.com/ssltest/analyze.html?d=mixpanel.com&s=67.228.186.42&hideResults=on

**Comment by hamann** [comment]
I can reproduce this error if a server is providing a root certificate. We tested it with our wildcard certificate and included [this bundle](https://knowledge.rapidssl.com/support/ssl-certificate-support/index?page=content&actp=CROSSLINK&id=SO26459)

**Comment by benoitc** [comment]
@hamann what do you mean by `a server is providing a root certificate` exactly? Trying to figure what I should fix at the end. The error looks similar to #196 for me. There should eb an intermediary step in hackney to validate if we have the whole chain or not. something like it.

**Comment by hamann** [comment]
@benoitc I mean when the server is configured to send the main, intermediate and root certificate. Afaik the root cert isn't necessary and should be ignored by clients, see [1], but some servers are doing it anyway, like https://data.mixpanel.com, or https://track.gaug.es for which hackney is failing, too.

Please have a look at the results from ssllabs.com as posted above and compare the certificates path for mixpanel.com with data.mixpanel.com. Maybe that is explaining it better.

[1] https://community.qualys.com/thread/11026

**Comment by hamann** [comment]
Some update to this...
I'm no SSL expert but I think know what's happening here. 

```sh 
$ openssl s_client -showcerts -connect data.mixpanel.com:443
```
This reports three certificates. The main, ending on `CLTABcU=`, the intermediate (`mJMcCaY`) and a root cert `7mHyhD8S`.

I've put the reported root cert in a file to get the fingerprint:
```sh
$ openssl x509 -noout -in ca2.cert -fingerprint -sha1
SHA1 Fingerprint=73:59:75:5C:6D:F9:A0:AB:C3:06:0B:CE:36:95:64:C8:EC:45:42:A3
```
, opened a browser to get the sha1 fingerprint for 'Geotrust Global CA' => 'DE 28 F4 A4 FF E5 B9 2F A3 C5 03 D1 A3 49 A7 F9 96 2A 82 12'. 
Then I googled for that fingerprint which lead me to  https://www.geotrust.com/resources/root-certificates/ and included the cert under https://www.geotrust.com/resources/root_certificates/certificates/GeoTrust_Global_CA.pem in one of our staging server (we’re also using RapidSSL wildcard certs) and tada => hackney is working.

So me assumption is that data.mixpanel.com _and_ track.gaug.es are sending the wrong root certificate (because probably [RapidSSL](https://knowledge.rapidssl.com/support/ssl-certificate-support/index?page=content&actp=CROSSLINK&id=SO26459) attached the wrong root cert `HyhD8S` on their page?).
If that's the case, than probably hackney does the right thing, it fails. Or does it the wrong thing, because it behaves differently than other clients/browsers?

Can someone confirm that?

**Comment by benoitc** [comment]
fixed via 99434ee79076c014945e0785101c0d229515fb78 . Thanks for all the feedback!
